jax/experimental/key_reuse/_forwarding.py

```diff
@@ -303,3 +304,21 @@ def _while_key_type_signature(eqn, args_consumed):
   return body_signature
 
 key_reuse_signatures_dynamic[jax.lax.while_p] = _while_key_type_signature
+
+def _remat_key_type_signature(eqn, args_consumed):
+  jaxpr = eqn.params['jaxpr']
+  forwarded_inputs = {i: eqn.invars.index(var) for i, var in enumerate(eqn.invars)
+                      if var in eqn.invars[:i]}
+  if eqn.params['differentiated']:
+    # When differentiated, the contents of the jaxpr have already been executed
+    # and so we need not check for reuse. But forwards are still relevant:
+    sig = get_jaxpr_type_signature(jaxpr, consumed_inputs=None,
+                                   forwarded_inputs=forwarded_inputs)
```

Comment:
This isn't directly related to the PR at hand but: I was a bit surprised that `get_jaxpr_type_signature` needs to take `consumed_inputs` and `forwarded_inputs` as arguments. If the function is just computing the rng type signature of the jaxpr it's given, then this seems to make the type dependent on the calling context. If the idea is that passing in this information lets us raise errors from within the jaxpr related to reuse of a caller's key, then it's not clear to me that that's preferable to raising the error at the jaxpr call site. That is, it seems we could instead have a function like `get_jaxpr_type_signature(jaxpr: Jaxpr) -> KeyReuseSignatureWithForwards`. What do you think?

Reply:
The reason for `consumed_inputs` is to raise better errors when a consumed input is re-consumed in the jaxpr. For example, say there's a 1000 line jaxpr that has a signature `{Sink(0)}`. We could trace it without the context, return this signature, and then say "sorry, input `0` is already consumed" and leave it up to the user to read through the 1000 lines to figure out where it's happening. But by passing `consumed_inputs` to the interpreter, we can instead tell the user exactly where in the jaxpr that reuse is happening.

`forwarded_inputs` is related: say you do something like this:
```python
key = jax.random.key(0)
func(key, key, key, key)
```
If the function consumes more than one of its inputs, this is key reuse. Again, we could just parse the function without that context and at the end tell the user that somewhere in the 1000-line jaxpr an input has been consumed twice, or we can trace it with that context and instead tell them exactly where in the jaxpr it went wrong.

In both cases I've opted to use the context to improve the error message.

Reply:
> then it's not clear to me that that's preferable to raising the error at the jaxpr call site

Note that the entire logic is implemented via `check_jaxpr`, so if we remove this contextual information, then the ultimate "call site" is the top-level function that you've wrapped in `jit`.

Reply:
Updated after offline convo!